### DISCUSSION TITLE
Partial update nested dictionary

### DISCUSSION THREAD

**Discussion by Nemesis7** [sqlalchemy/sqlalchemy#9409] [Usage Questions]
Hi,

I want to be able to do a partial update on a JSONB field instead of updating the whole field. My (simplified) setup:

**table**
```python
class Task(Base):
    __tablename__ = "tasks"

form = Column(MutableDict.as_mutable(JSONB))
```
I've used the MutableDict implementation proposed by @zzzeek at https://github.com/sqlalchemy/sqlalchemy/discussions/6473

**update query**
```python
statement = update(self.table).where(getattr(self.table, self.pk) == id).values(data.dict())
await session.execute(statement)
await session.commit()
```

Note that the data is a Pydantic model that has `exclude_unset = True`

**model**
```python
class Attributes(BaseModel):
   attribute_1: Optional[str]
   attribute_2: Optional[str]

class Form(BaseModel):
    domain: Optional[Domain]
    action: Optional[str]
    attributes: Attributes
    current_details: CustomerDetails
    new_details: CustomerDetails
```

**payload**
```json
{
        "domain": "SomeDomainName",
        "attributes": {
            "attribute_1": "value_1",
            "attribute_2": "value_2"
        },
        "action": "ACTION",
        "current_details": {},
        "new_details": {}
}
```

When I initially create the entry, I provide values for both attributes, but when I want to update it, it nullifies the attribute I didn't provide, even though it's not being sent in the data.dict() result. What am I doing wrong?

**Comment by zzzeek** [comment]
that dictionary is just one monolithic value to SQLAlchemy.   In the operation you're running, you're instructing it to emit a single UPDATE statement directly.   This looks like `UPDATE table SET my_json_field=<data>`, that's it.  There's no indication of whatever data was there previously.

To partially update the fields, the most straightforward way is to SELECT this data first, get the dictionary you want to update it, modify it on the Python side, then send it out again.   Using ORM unit of work patterns instead of update() this would be pretty easy, or you can just do SELECT and pass the completed dictionary back to your UPDATE.

OTOH, if you are going for a performance concern where you don't want to SELECT that data at all, you need to use special database features, which are only on PostgreSQL, in order to emit an UPDATE that only modifies part of the dictionary.   This syntax is not as easy to use and I would only use it if you are looking to UPDATE lots of rows very quickly, or which have enormous JSON fields that you don't want to pull over the network first.   The syntax is the [jsonb_set()](https://stackoverflow.com/a/35349699/34549) SQL function on PostgreSQL and it would look like:


    statement = update(self.table).where(getattr(self.table, self.pk) == id).values(data=func.jsonb_set(self.table.c.data, "{foo: bar}"))

**Reply by Nemesis7** [reply]
Thanks @zzzeek, I am indeed on PostgreSQL and I tried using the jsonb_set(), but could not make it work. When I try to do this:
```python
                statement = (
                    update(self.table)
                    .where(getattr(self.table, self.pk) == id)
                    .values(
                        **data.dict(exclude={"form"}),
                        form=func.jsonb_set(self.table.form, "{}", json.dumps(data.form)),
                    )
                )
```

next to the form field, I have other fields in the table that I need to be able to update, hence the split between data without the form attribute vs. the mutable form field.

I get this error:
```
sqlalchemy.exc.ProgrammingError: (sqlalchemy.dialects.postgresql.asyncpg.ProgrammingError) <class 'asyncpg.exceptions.UndefinedFunctionError'>: function jsonb_set(jsonb, character varying, boolean) does not exist
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
```

Also tried:
`form=func.jsonb_set(self.table.form  + cast(data.form, JSONB))`
`form=func.jsonb_set(self.table.form, cast(data.form, JSONB))`
`form=func.jsonb_set(cast(self.table.form, JSONB) + cast(data.form, JSONB))`

**Reply by zzzeek** [reply]
the casts should do it, so try to work out the raw SQL that PostgreSQL accepts and I will show you how to duplicate it

**Reply by Nemesis7** [reply]
It's should be this `UPDATE tasks SET tasks.form = tasks.form || '{"attributes":{"attribute_1":"value_3"}' WHERE tasks.id = id;` , found at: https://github.com/SaraAlhaddadi/jsonb_posts/blob/main/update_and_delete.sql#L137. That corresponds to the first one from the "also tried" attempts. However, that doesn't really merges the provided dict/json with the existing one, but it overwrites the existing data it seems.

`form=self.table.form + cast(data.form, JSONB)` should be doing the same as the one with the jsonb_set(), however it seems that the function is not called correctly in my attempts.

**Reply by anirudhwebbai** [reply]
@Nemesis7 were you able to fix this? I also ran into same issue.

**Reply by davidrios** [reply]
Note that in addition to the aforementioned performance issue with large JSON fields, by doing the proposed select -> modify locally -> update, one is introducing a much more serious risk of race condition-induced data loss. As a general rule one should  **never** do an update based on a value that was selected previously, as the value could have change between the select and update operations. Either do an atomic operation or lock the row in a transaction beforehand. Note that doing a select **does not** lock the row.

`jsonb_set` only sets a single path (overwriting it) and only works if the path already exists, so it'll not solve OP's issue.

To solve OP's problem I can think of two approaches:

1. Using a single atomic operation entirely in the database
2. Merging the dicts locally, while being careful to avoid race conditions


### 1. Using a single operation

I have implemented this recursive Python function that generates a single update operation that recursively merges the new dict to the current one in the database:

```python
import itertools
from sqlalchemy import func, sql
from sqlalchemy.dialects.postgresql import JSONB

def jsonb_deep_merge(col, value):
    """
    Generates a query to recursively merge `value` with the JSONB column `col`.

    Only dicts are merged, other types are replaced. `value` must be JSON serializable.
    """

    if not isinstance(value, dict):
        return func.cast(value, JSONB)

    return sql.case(
        (func.jsonb_typeof(col) == "object", col), else_=func.cast({}, JSONB)
    ).concat(
        func.jsonb_build_object(
            *itertools.chain.from_iterable(
                (k, jsonb_deep_merge(col[k], v)) for k, v in value.items()
            )
        )
    )
```

Note that it doesn't do anything locally, this is necessary for the operation to be atomic.

Then OP could use it like so:

```python
statement = (
    update(self.table)
    .where(getattr(self.table, self.pk) == id)
    .values(
        **data.dict(exclude={"form"}),
        form=jsonb_deep_merge(self.table.form, data.form), # maybe data.form.dict()
    )
)
```


### 2. Merging the dicts locally

The dicts can also be merged locally in Python, with the caveat that it might be inefficient for large JSON columns, and being careful to avoid a race condition in the database:

```python
# start a transaction. can be skipped if you're ABSOLUTELY SURE one is already started
session.begin()

pk_col = getattr(self.table, self.pk)
where_filter = pk_col == id

# do a dummy update to lock the row
statement = update(self.table).where(where_filter).values({pk_col: pk_col})
session.execute(statement)

# the row is locked in the transaction, so now it's safe to read the value
db_form_value = session.scalar(select(self.table.c.form).where(where_filter))

# Now you need to recursively merge the new dict to the old one.
# I'm assuming a function that does that already exists and is called mergedeep
merged_form = mergedeep(db_form_value or {}, new_form)

# now do the real update
statement = (
    update(self.table)
    .where(where_filter)
    .values(
        **data.dict(exclude={"form"}),
        form=merged_form,
    )
)
session.execute(statement)

# eventually commit the transaction
session.commit()
```

There're a few different approaches for implementing deep merging of dicts in Python, look one up on Google that fits your requirements.

BTW, sorry for necroing this, but it's one of the first Google results on the issue.

**Reply by CaselIT** [reply]
> As a general rule one should never do an update based on a value that was selected previously, as the value could have change between the select and update operation

You can issue a "for-update" lock in a select https://docs.sqlalchemy.org/en/20/core/selectable.html#sqlalchemy.sql.expression.Select.with_for_update. There is no need of "# do a dummy update to lock the row"

Your other point regarding performance stand though
